src/app/reducers/references.handlers.ts

```diff
@@ -2,71 +2,87 @@ import { Node as ProsemirrorNode } from 'prosemirror-model';
 
 import { ManuscriptHistoryState } from 'app/store';
 import { getRefNodeText, Reference, sortReferencesList } from 'app/models/reference';
-import { cloneManuscript } from 'app/utils/state.utils';
-import { createDiff, updateManuscriptState } from 'app/utils/history.utils';
+import { UpdateObjectChange } from 'app/utils/history/update-object-change';
+import { ProsemirrorChange } from 'app/utils/history/prosemirror-change';
+import { BatchChange } from 'app/utils/history/change';
+import { AddObjectChange } from 'app/utils/history/add-object-change';
+import { RearrangingChange } from 'app/utils/history/rearranging-change';
+import { DeleteObjectChange } from 'app/utils/history/delete-object-change';
 
 export function updateReference(state: ManuscriptHistoryState, payload: Reference): ManuscriptHistoryState {
   const referenceIndex = state.data.present.references.findIndex(({ id }) => id === payload.id);
+  const referenceChange = new UpdateObjectChange(
+    `references.${referenceIndex}`,
+    state.data.present.references[referenceIndex],
+    payload
+  );
+
   const { body } = state.data.present;
-  const changes = body.tr;
+  const transaction = body.tr;
   const newAttrs = { refId: payload.id, refText: getRefNodeText(payload) };
 
   body.doc.descendants((node: ProsemirrorNode, pos: number, parent: ProsemirrorNode) => {
     if (node.type.name === 'refCitation' && node.attrs['refId'] === payload.id) {
-      changes.replaceWith(pos, pos + node.nodeSize, body.schema.nodes['refCitation'].create(newAttrs));
+      transaction.replaceWith(pos, pos + node.nodeSize, body.schema.nodes['refCitation'].create(newAttrs));
     }
     return Boolean(node.childCount);
   });
-  const newState = updateManuscriptState(state.data, 'body', changes);
-  newState.present.references[referenceIndex] = payload;
-  sortReferencesList(newState.present.references);
-  newState.past[newState.past.length - 1]['references'] = state.data.present.references;
+
+  const change = new BatchChange([referenceChange, new ProsemirrorChange('body', transaction)]);
 
   return {
     ...state,
-    data: newState
+    data: {
+      past: [...state.data.past, change],
+      present: change.applyChange(state.data.present),
+      future: []
+    }
   };
 }
 
 export function addReference(state: ManuscriptHistoryState, payload: Reference): ManuscriptHistoryState {
-  const newDiff = createDiff({
-    references: state.data.present.references
-  });
+  const addReferenceChange = new AddObjectChange('references', payload, 'id');
+  const updatedManuscript = addReferenceChange.applyChange(state.data.present);
+  const updatedReferenceList = [...updatedManuscript.references];
+  sortReferencesList(updatedReferenceList);
 
-  const newManuscript = cloneManuscript(state.data.present);
-  newManuscript.references.push(payload);
-  sortReferencesList(newManuscript.references);
+  const sortReferencesChange = RearrangingChange.createFromListRearrange(
```

Comment:
Also, adding a new reference doesn't trigger a sort

Reply:
This is still an issue, but its the same in staging hence not a regression. I'll raise a new issue to track this.